Incorrect identically-labeled DataFrame objects Exception comes with df.compare method #50083

tpackard1 opened the pull request:
- [ ] closes #50083
- [ ] All code checks passed

tpackard1 commented:
First time contributing to the code base and hopefully everything goes smoothly.

MarcoGorelli reviewed (changes requested):
this is a nice start! could you also:

- fixup the failing tests
- add a whatsnew note

tpackard1 commented:
I will try to fix the failing tests tonight.

tpackard1 commented:
@MarcoGorelli  would the following work as a whatsnew note?
**Whatsnew Note**

- Updated the Exception messages for df.compare method to include that both index and columns that are identically-labeled can be compared.
- Updated the test associated with these Exceptions.

MarcoGorelli commented:
You can just say that you improved the error message - I think there's some similar entries in the what's new file

No need to mention tests, it's a user facing note

tpackard1 commented:
@MarcoGorelli I have updated the whats new file. Please review whenever you have a moment

MarcoGorelli commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -97,6 +97,7 @@ Other enhancements
 - Added :meth:`Index.infer_objects` analogous to :meth:`Series.infer_objects` (:issue:`50034`)
 - Added ``copy`` parameter to :meth:`Series.infer_objects` and :meth:`DataFrame.infer_objects`, passing ``False`` will avoid making copies for series or columns that are already non-object or where no better dtype can be inferred (:issue:`50096`)
 - :meth:`DataFrame.plot.hist` now recognizes ``xlabel`` and ``ylabel`` arguments (:issue:`49793`)
+- :func:`DataFrame.compare` exception message thrown improved to include that both index and columns that are identically labeled can be compared. (:issue:`50083`)
```
```suggestion
- Improved error message in :func:`DataFrame.compare` to clarify that "identically labelled" refers to both index and columns (:issue:`50083`)
```

tpackard1 replied:
I updated the whatsnew note to match your comment above

MarcoGorelli reviewed (changes requested):
Looks good, just got a minor comment on the whatsnew note

MarcoGorelli reviewed (changes requested):
Wait, something went wrong when merging/rebasing I presume, you've lost all your changes, all I can see if the whatsnew note change

tpackard1 committed:
added back in changes

tpackard1 committed:
updated whatsnew note

tpackard1 commented:
I have updated the whatsnew comment and added back in the all the other changes. I wanted to make sure all the checks were passed before requesting a review but some of the checks keep failing with the following error message `ERROR  - SystemError: initialization of _internal failed without raising an exception`. Is this due to the change requested? Also sorry this is taking so much or your time @MarcoGorelli  but I appreciate your time as this process as taught me alot.

MarcoGorelli commented:
Hey - CI failures are unrelated, they're showing up in other PRs, I'll see what's going on there

tpackard1 committed:
Merge remote-tracking branch 'upstream/main' into test-build-test

MarcoGorelli commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -98,6 +98,7 @@ Other enhancements
 - Added :meth:`Index.infer_objects` analogous to :meth:`Series.infer_objects` (:issue:`50034`)
 - Added ``copy`` parameter to :meth:`Series.infer_objects` and :meth:`DataFrame.infer_objects`, passing ``False`` will avoid making copies for series or columns that are already non-object or where no better dtype can be inferred (:issue:`50096`)
 - :meth:`DataFrame.plot.hist` now recognizes ``xlabel`` and ``ylabel`` arguments (:issue:`49793`)
+- Improved error message in :func:`DataFrame.compare` to clarify that "identically labelled" refers to both index and columns (:issue:`50083`)
```
sorry, just noticed that this affects much more than just `compare`, e.g:
```python
In [8]: df == dfother
---------------------------------------------------------------------------
ValueError: Can only compare identically-labeled DataFrame objects
```
would also have thrown this error

let's rephrase as
```python
Improved error message when trying to align :class:`DataFrame` objects (for example, in :func:`DataFrame.compare`) to clarify ...
```

MarcoGorelli reviewed (changes requested):
one last minor comment, sorry - then I think it's good to go 🚀

tpackard1 committed:
reworded the whatsnew comment

tpackard1 committed:
Merge remote-tracking branch 'upstream/main' into test-build-test

MarcoGorelli reviewed (approved):
Looks good to me, thanks @tpackard1

mroeschke commented:
Thanks @tpackard1
